[async-mqtt] Add new port

redboltz opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/ -->

<!-- If this PR updates an existing port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [ ] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is added to each modified port's versions file.

END OF PORT UPDATE CHECKLIST (delete this line) -->

If this PR adds a new port, please uncomment and fill out this checklist:

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line)

redboltz committed:
[async-mqtt] Add async-mqtt port.

redboltz commented:
@microsoft-github-policy-service agree

FrankXie05 commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
+
+file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
```suggestion
```
Just need function `vcpkg_install_copyright`.

FrankXie05 commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO redboltz/async_mqtt
+    REF 1.0.1
```
```suggestion
    REF "${VERSION}"
```
We can use the new function.

FrankXie05 commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(PACKAGE_NAME async_mqtt_iface CONFIG_PATH lib/cmake/async_mqtt_iface)
```
```suggestion
vcpkg_cmake_config_fixup(PACKAGE_NAME async_mqtt_iface CONFIG_PATH "lib/cmake/async_mqtt_iface")
```

FrankXie05 commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+
+vcpkg_cmake_config_fixup(PACKAGE_NAME async_mqtt_iface CONFIG_PATH lib/cmake/async_mqtt_iface)
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug" "${CURRENT_PACKAGES_DIR}/lib")
```
Is this port a `header-only` or only `dynamic`.?

redboltz replied:
This portis a header-only.

redboltz replied:
Should these two lines be removed if the port is a header-only library?

FrankXie05 replied:
No, keep two lines. We don't need empty folders.

FrankXie05 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". 
          That way, I can be aware that you've responded since you can't modify the tags.

FrankXie05 commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+vcpkg_from_github(
```
```suggestion
#header-only library

vcpkg_from_github(
```

redboltz replied:
I looked over this comment. I just reflected it.

dg0yt commented on ports/async-mqtt/portfile.cmake:
```diff
@@ -0,0 +1,27 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+    -DASYNC_MQTT_BUILD_TOOLS=OFF
+    -DASYNC_MQTT_BUILD_EXAMPLES=OFF
+    -DASYNC_MQTT_BUILD_UNIT_TESTS=OFF
+    -DASYNC_MQTT_BUILD_SYSTEM_TESTS=OFF
+    -DCMAKE_DISABLE_FIND_PACKAGE_Doxygen=ON
```
Please indent the actual options:
~~~
    OPTIONS
        -DASYNC_MQTT_BUILD_TOOLS=OFF
        -DASYNC_MQTT_BUILD_EXAMPLES=OFF
        -DASYNC_MQTT_BUILD_UNIT_TESTS=OFF
        -DASYNC_MQTT_BUILD_SYSTEM_TESTS=OFF
        -DCMAKE_DISABLE_FIND_PACKAGE_Doxygen=ON
~~~

redboltz committed:
Reflect review comments.

Add header-only on the top of file.
Use ${VERSION} instead of the literal.
Fix indent.
Add double qoute to path
Remove copyright install.

redboltz commented:
I think that all changes request is reflected.

redboltz committed:
Set build type to release.

redboltz commented:
https://github.com/microsoft/vcpkg/pull/31902#discussion_r1224148906 is fixed too.
So now ready to review.

dg0yt reviewed (approved):
Just from looking at it in GH, only one nit-pick left.

The port is submitted from the owner, and it is not yet a project with many GH stars, so vcpkg maintainers must decide if they want the port under this name or under `redboltz-async-mqtt`. (There is at least a JS project under the same name. https://github.com/search?q=async-mqtt&type=repositories)

redboltz committed:
Removed debug directory removal.

redboltz commented:
> Just from looking at it in GH, only one nit-pick left.
> 
> The port is submitted from the owner, and it is not yet a project with many GH stars, so vcpkg maintainers must decide if they want the port under this name or under `redboltz-async-mqtt`. (There is at least a JS project under the same name. https://github.com/search?q=async-mqtt&type=repositories)

Thank you for the comment. I fixed that removing debug directory.
Let me explain some background. I am the author both 
https://github.com/redboltz/async_mqtt and
https://github.com/redboltz/mqtt_cpp .
mqtt_cpp has been registered in vcpkg. Some contributor did it.
async_mqtt is successor project of mqtt_cpp. It coveres all asynchronous functionality of mqtt_cpp, but using more elegant way.

After I released async_mqtt, vcpkg support is requested by user https://github.com/redboltz/async_mqtt/issues/16
So I createed this PR. I accept async-mqtt, redboltz-async-mqtt, or perhaps async-mqtt-cpp.
(The name of https://github.com/redboltz/async_mqtt is not changed.)

FrankXie05 commented:
@redboltz 
It seems that when generating the cmake configuration file, a parenthesis is missing.

`async_mqtt_ifaceConfig.cmake`:
```
set(Boost_USE_MULTITHREADED ON)
include(CMakeFindDependencyMacro)
find_dependency((Threads)
find_dependency((Boost 1.81.0 COMPONENTS )
...
...
```
```
CMake Error at E:/usage-test/vcpkg/installed/x64-windows/share/async_mqtt_iface/async_mqtt_ifaceConfig.cmake:3:
  Parse error.  Function missing ending ")".  End of file reached.
```

dg0yt commented:
> find_dependency((Boost 1.81.0 COMPONENTS )

There is a double `(`.

FrankXie05 commented:
```
find_dependency((Threads)
```

FrankXie05 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". 
          That way, I can be aware that you've responded since you can't modify the tags.

redboltz committed:
Fix async_mqtt installed cmake config.

redboltz commented:
I fixed async_mqtt cmake config issue and then released it as 1.0.2.
Then I updated this PR.

redboltz commented:
All checks have passed, now ready for review.

FrankXie05 commented:
The usage has been tested successfully locally.
```
async-mqtt provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(async_mqtt_iface CONFIG REQUIRED)
    target_link_libraries(main PRIVATE async_mqtt_iface::async_mqtt_iface)
```
